nsis/src/nsis/format/NsisScriptHeader.java

```diff
@@ -17,71 +17,80 @@
 	private int archiveSize;
 	private int compressedHeaderSize;
 	private int flags;
+	private static Structure STRUCTURE;
 
 	public NsisScriptHeader(BinaryReader reader) throws IOException {
-		setMagic(reader.readNextByteArray(NsisConstants.NSIS_MAGIC.length));
+		this.magic = reader.readNextByteArray(NsisConstants.NSIS_MAGIC.length);
 		if (!Arrays.equals(NsisConstants.NSIS_MAGIC, getMagic())) {
-			throw new IOException("not a nsis file.");
+			throw new IOException("Not a nsis file.");
 		}
 
-		setInflatedHeaderSize(reader.readNextInt());
-		setArchiveSize(reader.readNextInt());
-		setCompressedHeaderSize(reader.readNextInt());
-		setFlags(reader.readNextInt());
+		this.inflatedHeaderSize = reader.readNextInt();
+		this.archiveSize = reader.readNextInt();
+		this.compressedHeaderSize = reader.readNextInt();
+		this.flags = reader.readNextInt();
+		checkHeaderCompression(reader);
 	}
 
 	@Override
 	public DataType toDataType() throws DuplicateNameException, IOException {
-		Structure structure = new StructureDataType("script_header", 0);
-		structure.add(STRING, NsisConstants.NSIS_MAGIC.length, "magic", null);
-		structure.add(DWORD, 4, "inf_size", null);
-		structure.add(DWORD, 4, "hdr_size", null);
-		structure.add(DWORD, 4, "cmpr_hdr_size", null);
-		structure.add(DWORD, 4, "flags", null);
-		return structure;
+		if(STRUCTURE == null) {
+			STRUCTURE = new StructureDataType("script_header", 0);
+			STRUCTURE.add(STRING, NsisConstants.NSIS_MAGIC.length, "magic", null);
+			STRUCTURE.add(DWORD, DWORD.getLength(), "inf_size", null);
+			STRUCTURE.add(DWORD, DWORD.getLength(), "hdr_size", null);
+			STRUCTURE.add(DWORD, DWORD.getLength(), "cmpr_hdr_size", null);
+			STRUCTURE.add(DWORD, DWORD.getLength(), "flags", null);
+		}
+		return STRUCTURE;
 	}
 
 	public byte[] getMagic() {
 		return magic;
 	}
 
-	public void setMagic(byte[] magic) {
-		this.magic = magic;
-	}
-
 	public int getInflatedHeaderSize() {
 		return inflatedHeaderSize;
 	}
 
-	public void setInflatedHeaderSize(int inflated_header_size) {
-		this.inflatedHeaderSize = inflated_header_size;
-	}
-
 	public int getArchiveSize() {
 		return archiveSize;
 	}
 
-	public void setArchiveSize(int archiveSize) {
-		this.archiveSize = archiveSize;
-	}
-
 	public int getCompressedHeaderSize() {
 		return compressedHeaderSize;
 	}
 
-	public void setCompressedHeaderSize(int compressedHeaderSize) {
-		this.compressedHeaderSize = compressedHeaderSize;
-	}
-
 	public static int getHeaderSize() {
-		return NsisConstants.NSIS_MAGIC.length + 4 + 4 + 4 + 4;
+		return STRUCTURE.getLength();
 	}
 
 	public int getFlags() {
 		return flags;
 	}
 
-	public void setFlags(int flags) {
-		this.flags = flags;
+	public void checkHeaderCompression(BinaryReader reader) {
+		if ((this.compressedHeaderSize & 0x80000000) == 0) {
+			System.out.print("Header is not compressed!\n");
```

Comment:
This function should signal somehow that the check was unsuccessful, either by returning a value to be checked by the caller or by throwing exceptions, like the proposed above `nsis.format.InvalidFormat`.

Right now it only prints error messages, but otherwise the errors are effectively swallowed.

Reply:
This function is broken (#9 ), I only changed its location because at first it was in NsisLoader.java. I was thinking of addressing this function in the next PR, including error handling. Does that make sense?

Reply:
Done